pkg/peertls/extensions.go

```
@@ -20,59 +69,330 @@ type TLSExtConfig struct {
+func NewRevocationDBRedis(address string) (*RevocationDB, error) {
+	client, err := redis.NewClientFrom(address)
+	if err != nil {
+		return nil, err
+	}
+	return &RevocationDB{
+		DB: client,
+	}, nil
+}
+
+// NewRevocationExt generates a revocation extension for a certificate.
+func NewRevocationExt(key crypto.PrivateKey, revokedCert *x509.Certificate) (pkix.Extension, error) {
+	nowUnix := time.Now().Unix()
+
+	hash, err := hashBytes(revokedCert.Raw)
+	if err != nil {
+		return pkix.Extension{}, err
+	}
+	rev := Revocation{
+		Timestamp: nowUnix,
+		CertHash:  make([]byte, len(hash)),
+	}
+	copy(rev.CertHash, hash)
+
+	if err := rev.Sign(key); err != nil {
+		return pkix.Extension{}, err
+	}
+
+	revBytes, err := rev.Marshal()
+	if err != nil {
+		return pkix.Extension{}, err
+	}
+
+	ext := pkix.Extension{
+		Id:    ExtensionIDs[RevocationExtID],
+		Value: make([]byte, len(revBytes)),
+	}
+	copy(ext.Value, revBytes)
+
+	return ext, nil
+}
+
+// AddRevocationExt generates a revocation extension for a cert and attaches it
+// to the cert which will replace the revoked cert.
+func AddRevocationExt(key crypto.PrivateKey, revokedCert, newCert *x509.Certificate) error {
+	ext, err := NewRevocationExt(key, revokedCert)
+	if err != nil {
+		return err
+	}
+	err = AddExtension(newCert, ext)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// AddSignedCertExt generates a signed certificate extension for a cert and attaches
+// it to that cert.
+func AddSignedCertExt(key crypto.PrivateKey, cert *x509.Certificate) error {
+	signature, err := signHashOf(key, cert.RawTBSCertificate)
+	if err != nil {
+		return err
+	}
+
+	err = AddExtension(cert, pkix.Extension{
+		Id:    ExtensionIDs[SignedCertExtID],
+		Value: signature,
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// AddExtension adds one or more extensions to a certificate
+func AddExtension(cert *x509.Certificate, exts ...pkix.Extension) (err error) {
+	if len(exts) == 0 {
+		return nil
```

curious, would you want to return an error?

---

:man_shrugging: